chain/client/src/stateless_validation/chunk_endorsement_tracker.rs

```diff
@@ -40,33 +54,64 @@ impl ChunkEndorsementTracker {
+            // We can use a different cache size if needed, it does not have to be the same as for `chunk_endorsements`.
+            pending_chunk_endorsements: SyncLruCache::new(NUM_CHUNKS_IN_CHUNK_ENDORSEMENTS_CACHE),
         }
     }
 
+    pub fn process_pending_endorsements(
+        &self,
+        chunk_header: &ShardChunkHeader,
+    ) -> Result<(), Error> {
+        let chunk_hash = &chunk_header.chunk_hash();
+        let chunk_endorsements = {
+            let mut guard = self.pending_chunk_endorsements.lock();
+            guard.pop(chunk_hash)
+        };
+        let chunk_endorsements = match chunk_endorsements {
+            Some(chunk_endorsements) => chunk_endorsements,
+            None => {
+                tracing::debug!(target: "stateless_validation", ?chunk_hash, "No pending chunk endorsements.");
+                return Ok(());
+            }
+        };
+        chunk_endorsements.values().try_for_each(|endorsement| {
+            self.process_chunk_endorsement(endorsement.clone(), Some(&chunk_header))
+        })
+    }
+
     /// Function to process an incoming chunk endorsement from chunk validators.
-    /// We first verify the chunk endorsement and then store it in a cache.
+    /// If the chunk header is available, we will verify the chunk endorsement and then store it in a cache.
+    /// Otherwise, we store the endorsement in a separate cache of endorsements to be processed when the chunk is ready.
     /// We would later include the endorsements in the block production.
     pub(crate) fn process_chunk_endorsement(
         &self,
-        chunk_header: &ShardChunkHeader,
         endorsement: ChunkEndorsement,
+        chunk_header: Option<&ShardChunkHeader>,
     ) -> Result<(), Error> {
         let chunk_hash = endorsement.chunk_hash();
         let account_id = &endorsement.account_id;
 
+        let endorsement_cache = if chunk_header.is_some() {
+            &self.chunk_endorsements
+        } else {
+            &self.pending_chunk_endorsements
+        };
+
         // If we have already processed this chunk endorsement, return early.
-        if self
-            .chunk_endorsements
+        if endorsement_cache
             .get(chunk_hash)
             .is_some_and(|existing_endorsements| existing_endorsements.get(account_id).is_some())
         {
             tracing::debug!(target: "stateless_validation", ?endorsement, "Already received chunk endorsement.");
             return Ok(());
         }
 
-        if !self.epoch_manager.verify_chunk_endorsement(chunk_header, &endorsement)? {
-            tracing::error!(target: "stateless_validation", ?endorsement, "Invalid chunk endorsement.");
-            return Err(Error::InvalidChunkEndorsement);
+        if let Some(chunk_header) = chunk_header {
+            if !self.epoch_manager.verify_chunk_endorsement(&chunk_header, &endorsement)? {
+                tracing::error!(target: "stateless_validation", ?endorsement, "Invalid chunk endorsement.");
```

Comment:
ditto - add the chunk hash

Reply:
As above, it is already part of `?endorsement`.